### ISSUE TITLE
Ordered Effects

### ISSUE THREAD

**Issue by mike-thompson-day8** [day8/re-frame#639]
If I had my time over again, I would make effects be a seq, rather than a map.  The resulting effects DSL would be more expressive, at the cost of nesting. 

I have previously teased out the issues a bit here:
https://gist.github.com/mike-thompson-day8/1388e68391f4c6c1373c9555c08c7114#using-data-litterals

Advantages: 
  1. It gives a way to order effects (not that useful, but an interesting property)
  2. no need for `dispatch-n` , just use `:dispatch` multiple times (this same `-n` issues comes up a lot with other effects like http GETs etc)

Disadvantages:
  3. slightly noisier, visually, with the additional structural nesting
  4. is a breaking change 

It might be possible to work around the breaking change problem by introducing a new `reg-event-fx2` which expects the alternative kind of return. Leaving the old `reg-event-fx` alone.  The trouble is we'd end up with interceptors which worked with one kind of effects, but not the other. 

@olivergeorge has made an interesting suggestion which fixes the "breaking change" side of things, at the expense of a tiny bit of boilerplate.   He suggests introducing a  new `:fx` effect, which can take seq of other effects.  This `:fx` can be used in existing `reg-event-fx`:

```clj 
(reg-event-fx 
   :token 
   (fn [{:keys [db]} [_ id]]
      {:db    (assoc db  :something id)
       :fx     [{:dispatch [:blah]}
                 {:dispatch [:other]}
                 {:http  {:url   "http://abc.com/endpont  :method "GET"}}]}))
```

Notice the new `:fx` effect for which a `seq` of other effects is provided. 

***Question***: if done this way, what is the ordering between `:db` and those in `:fx`? `:db` first? 

***Question***:  I regard `:db` to be a normal effect (it is a side effect on `app-db`), so should it be put into `:fx` too.  (What happens if it is not, what is the ordering?)

```clj 
(reg-event-fx 
   :token 
   (fn [{:keys [db]} [_ id]]
      {:fx     [{:db    (assoc db  :something id)}
                {:dispatch [:blah]}
                 {:dispatch [:other]}
                 {:http  {:url   "http://abc.com/endpont  :method "GET"}}]}))
```

***Question***:  maybe just pairs like this

```clj 
(reg-event-fx 
   :token 
   (fn [{:keys [db]} [_ id]]
      {:fx     [:db    (assoc db  :something id)   ;; <-- not a map, just a pair
                :dispatch [:blah]
                :dispatch [:other]
                :http  {:url   "http://abc.com/endpont  :method "GET"}]}))
```

**Comment by olivergeorge** [comment]
Thanks for putting this together.

I vote this form
```cljs
(reg-event-fx 
   :token 
   (fn [{:keys [db]} [_ id]]
      {:db    (assoc db  :something id)
       :fx    [{:dispatch [:blah]}                             ;; <-- :fx takes a seq of effects
               {:dispatch [:other]}
               {:http  {:url   "http://abc.com/endpont  :method "GET"}}]}))
```

My argument being
* :db is special.  It's old-state (as arg) and new-state (as ret value).  It's implemented via reg-fx is really an implementation detail)
* this form provides a way cofx can flow through composed handlers with :db and :fx being special.
* this style is symmetrical (`arg` and `ret` value) allowing for composition `(comp (set-path [...]) (validate...)` or via `->`

The composition idea is a huge unlock for re-frame.  The handler calling convention complicates reuse.

e.g.
```
(defn set-foo [m] 
  (assoc-in m [:db :foo] (get-in m [:event 1]))

(defn do-bah [m] 
  (update-in m [:fx] conj {:bah x})
```

The _nature_ of db is the interesting insight in this.

PS.  I'd prefer :fx to expect [{:bah 1} ...] over [:bah 1 ...] for clarity/composability reasons.

**Comment by danielytics** [comment]
Regarding `db` being special, would the following be valid?
```
(reg-event-fx 
   :token 
   (fn [{:keys [db]} [_ id]]
      {:db  (assoc db  :something id)
       :dispatch  [:foo]                         ;; <-- using an "old-style" effect
       :fx  [:dispatch [:blah]
             :dispatch [:other]
             :http     {:url   "http://abc.com/endpont  :method "GET"}]}))
```
If `yes` (it is valid), then DB isn't really treated as special and it should probably be ok to have multiple `db` effects in the `fx` vector too. In this case, its important to define the order (I would say, all old-style non-`:fx` effects first, in arbitrary order, followed by the new-style `:fx` effects, in the specified order).

If `no` (it is not valid), then its not really backward compatible at all, in which case I'd suggest just dropping the outer map and returning the `fx` vector of pairs directly.

**Comment by olivergeorge** [comment]
Using a `(ref-fx :fx ...)` would be backward compatible (technically) and `:db`, like `:dispatch` could work in both places.  Problem is that if `:db` appears twice it would cause a conflict which clobbers one of the changes.

I guess a fix for that might be to keep a list of :fx which can appear only once and exclude them from use in the `(ref-fx :fx ...)` form.

**Comment by mike-thompson-day8** [comment]
I've let this idea marinate for a few of days and I'm yet to see any downsides. In fact, I'm getting more sure this is a great idea.  Even the key `:fx` works in nicely with `reg-fx` etc.

I will continue to let it sit for another couple of days just to see if any gremlins or improvements come to mind, before making it happen.  While it is only about 5 lines of code, it is all the docs changes which I'm not looking forward to - adding this feature will lead to a change to best practice.  I'll have to take on that task incrementally. 

The benefits:
  1. no breaking changes
  2. ordered effects
  3. the ability to automatically handle the `*-n` variant of all effects (eg: no need for `:dispatch-n`)
  4. a more natural way to compose event handlers from smaller functions. 

I'm currently favour this form:
```clj
(reg-event-fx 
   :token 
  (fn [{:keys [db]} [_ id]]
    {:db (assoc db :something id)
     :fx [{:dispatch  [:blah]}   
          (when seventies? {:hair "side burns"})     ;; nils are ignored
          {:http    {:method :GET  :url "http:/abc.com/endpoint"}}]}))
```

Notes:
  - the key `:fx` will be associated with a seq, typically a vector
  - the effects in `:fx` will be processed in order
  - each element in the seq will be a map which represents one effect. Eg `{:dispatch [:blah]}`
  - if an element of `:fx` is `nil`, it will be ignored, which allows effects to be conditionally included via a `when`
  - visually, this is more "noisy" than using the "pairs" format, and aesthetics do matter, but as you point out, this approach is easier to model mentally (teach and get right)
  - The `:db` effect will be processed before those in `:fx`.  In fact, `:db` will now be processed before all other effects, and `:fx` is just one of them. 

## Point 4 - Composing Event Handler 

Point 4 of the benefits list (above) is the most intriguing, and the main reason I'm still dwelling on it.  If we do this, I want to maximise the value we get from this approach. Now we have this idea in our head, and the smell of it in our nostrils, are there any other tweaks which would add even more value?

At a minimum, composing an event handler from smaller functions is now a lot more natural.  For example:

```clj
(reg-event-fx 
   :token 
  (fn [{:keys [db]} [_ id]]
   (-> {:db db 
        :fx  [{:dispatch [:token2]}]}                 ;; base structure
      (start-ui-throber)        ;; <-- might assoc into `:db`
      (GET "http://abc.com/endpoint")      ;; adds to `:fx`
      (POST "http://blah.com/endpoint")))  ;; adds to `:fx`

```

Notes:
  - in this imaginary example … 
  - the event handler starts by creating a data structure `{:db db :fx []}` which it then threads through the other functions, composing the effects. Except that isn't entirely true, because it adds a `:dispatch` effect using data literals at the time the structure is created.
  - `GET` and `POST` are imaginary functions which will conj an effect onto the `:fx` part of the supplied argument 
  - `start-ui-throbber` will probably `assoc` a value into some path of `:db` to cause the twirly thing to appear for the user

Compare the code above with the existing approach below which tends to encourage "effects via data literals". 

```clj
(reg-event-fx 
   :token 
  (fn [{:keys [db]} [_ id]]
   (-> {:db (assoc db :twirly true) 
        :dispatch [:blah]
        :http  (list 
                  {:method :GET :url "http://abc.com/endpoint" ...}
                  {:method :POST :url "http://blah.com/endpoint" ...})})))
```

There is something pleasingly explicit to me about maximising the use of data literals,  but composing effects via functions is also an interesting approach, and  I like the choice, and I like being able to mix the approaches as I see fit.  

## FSM Example

I am particularly happy because this new approach will make it easier to handle Finite State machines (FSM) within re-frame.  I look forward to being able to write: 
```clj
(reg-event-fx 
   :token 
  (fn [{:keys [db]} event]
   (-> {:db db :fx []}                 ;; base structure
       (fsm fsm-definition event)))    ;; process the finite state machine
```
where:
  - `fsm` is some general execution engine (function) which will step a definition forward. 
  - the new `:db` and `:fx` is just a lot more composable in the case that the FSM has to step through a couple of states before going async again. 


## Why `:fx` and `:db`

I still regard `:db` is just another side effect - it just involves a `reset!` of `app-db`.

But, we all intuit that `:db` is a different kind of side effect to those in `:fx`.  

But why? I'd enjoy getting a crisper explanation for this. 

It is something to do with `:fx`  effects being independent of each other and able to be done "later". But the`:db` side effect need to accumulate in situ because the accumulated results might be used by the next function. 

## reg-event-fx2


@olivergeorge I'm unclear on your `reg-event-fx2` proposal (duplicated below) 
```clj
(reg-event-fx2 :token (comp set-something fetch-data dispatch-other))
```

Are you suggesting a new registration function which sets up the `{:db db :fx []}` part automatically and then threads that through the supplied event handler (in your example created via `comp`?

**Comment by p-himik** [comment]
> adding this feature will lead to a change to best practice

But why? If someone doesn't care about the order (I find it to be the case in 90% of the situations I encounter), why suddenly writing
```
{:db (assoc db :a 1)
 :fx [{:dispatch [:do-stuff-to-a]}
      {:http {:method :get, :on-success [:inc-a], :on-failure [:dec-a]}}]}
```
is preferred to a simpler and yet as sufficient
```
{:db (assoc db :a 1)
 :dispatch [:do-stuff-to-a]
 :http {:method :get, :on-success [:inc-a], :on-failure [:dec-a]}}
```

**Comment by mike-thompson-day8** [comment]
@p-himik 
This really isn't about ordered effects.  (Even though the issue is titled that way). As you say, that's a minor usecase.

For one reason, it removes the need for all the `*-n` variants": there's both `:dispatch` and `dispatch-n`.  The same exists for `http` effects. 

Also, it allows for easier composing of effectfulness-causing functions.  This is the "Point 4" I talked about above.  If you have one function which wants to cause a dispatch, should it be adding to `:dispatch` or `:dispatch-n` ??  In the new scheme, always just conj onto `:fx`

**Comment by p-himik** [comment]
> it is better if there is only one recommended way to do things

Indeed. But I still don't really get why it has to be the `:fx` method.
And to be honest, I don't really get why there has to be a "recommended" way. I like having a set of tools, and I like to be able to choose the tool according to the task at hand and not because it was labeled as "recommended". We have at least 3 ways to get the first element of a vector that I can think of right away, and none of them is labeled as recommended.

Yes, `:fx` solves two pain points which one can encounter maybe once in 10-20 instances of using `ref-event-*` (just to mention it again - I can only speak from my own practice. Maybe there are people for whom the number is 3 or 100).
And I wouldn't whine were it the only thing that it does. Alas, `:fx` also introduces a trade-off - it makes things much more verbose in arguably one of the most frequent use cases, `:dispatch-n`:
```clj
{:dispatch-n [[:display-pop-up]
              [:start-throbber]
              [:fetch-data]
              [:start-countdown]]}
;; vs.
{:fx [{:dispatch [:display-pop-up]}
      {:dispatch [:start-throbber]}
      {:dispatch [:fetch-data]}
      {:dispatch [:start-countdown]}}
```
I definitely see such scenario _much_ more often than the need to compose or order the fx. And I personally would never recommend using `:fx` in this case if you don't know for sure that you will need to add some other effects in the middle of those events.

Actually, writing the above gave me an idea. What if `:fx` accepts a seq of not only maps but also of event vectors? In that case, the above could be written as
```clj
{:fx [[:display-pop-up]
      [:start-throbber]
      [:fetch-data]
      [:start-countdown]]}
```
It as all the benefits of the initially proposed version of `:fx` but it's also even less verbose (by the whopping 8 characters, yes) than `:dispatch-n`. I admit, it feels a bit "dirty". But at the same time it seems practical enough to justify that. Especially given that having either a vector or a map is somewhat a common practice with effects in general, e.g. as it's the case with `:http-xhrio`.

**Comment by mike-thompson-day8** [comment]
>  I don't really get why there has to be a "recommended" way.

The docs have to adopt a style. Best for newbies that way. 

> But I still don't really get why it has to be the `:fx` method.

I detailed the four benefits above. I also pointed out that, to me, the most interesting was "Point 4".

I'm confident that it will make it easier to implement certain aspects of FSM. For example, I can see my way clear to finishing this library: https://github.com/day8/re-frame-http-fx-alpha   (which stalled because I could never work out a clean way to compose state handlers)

But, hey, it is backwards compatible, so no need to change your existing.

**Comment by kirillsalykin** [comment]
What do you think about using vectors - so having the same structure as used in subs, I think it feels more consistent.

Something like this:
```
{:fx [[:dispatch [:display-pop-up]]
      [:dispatch [:start-throbber]]
      [:dispatch [:fetch-data]]
      [:dispatch [:start-countdown]]}
```

**Comment by mike-thompson-day8** [comment]
@ikitommi 
Evil thoughts can be very instructive and helpful :-)
I prefer my events to be pure data,  but it is interesting and instructive that your sketch is possible. 

@kirillsalykin
How would you propose to do an effect like this one: https://github.com/day8/re-frame-http-fx

**Comment by mike-thompson-day8** [comment]
Actions:
  - add the new `:fx` event handler 
    - [ ] add new `reg-fx` for `:fx`  (warnings if `:db` effects found)
    - [ ] add to API docs 
  - add guarantee that `:db` effect will be done first 
    - [ ] at the code level 
    - [ ] update tutorial docs in this area (there is a comment about no ordering somewhere, so remove that) 
    - [ ] update API docs to specify that `:db` is done first
  - [ ] update Changes docs 
  - [ ] release a new version 
  - [ ] review tutorial docs overall.  At what point do we introduce `:fx`? 

We'll defer introducing any new `reg-event-fx2` until we have more experience using this feature.

**Comment by kirillsalykin** [comment]
@mike-thompson-day8 

I imagine it like this:

```
{:fx [[:http-xhrio {:method          :post
                             :uri             "https://httpbin.org/post"
                             :params          data
                             :timeout         5000
                             :format          (ajax/json-request-format)
                             :response-format (ajax/json-response-format {:keywords? true})
                             :on-success      [::success-post-result]
                             :on-failure      [::failure-post-result]]}
```

Basic idea is to use vector literal instead of map literal - so it has some syntax as `reg-sub` 
Not something very important, just for consistency.

**Comment by kirillsalykin** [comment]
@mike-thompson-day8 
> If I had my time over again, I would make reg-event-fx return a seq of effects, rather than a map of effects. The resulting effects DSL would be more expressive, at the cost of some nesting.

Would it be possible to differentiate old behaviour from new based on type then?
eg if `reg-event-fx` returns map - it is old way fx, if vector/list - then the new way?

Thanks

**Comment by mike-thompson-day8** [comment]
@kirillsalykin

> Would it be possible to differentiate old behaviour from new based on type then?

Yes, it would be possible to differentiate.  But existing interceptors expect a map of effects. They would break.

**Comment by green-coder** [comment]
I suggest to stick with maps `{:dispatch [:display-pop-up]}` instead of vectors `[:dispatch [:display-pop-up]]`:

If one day you want to pass some meta data about the effect, the map would have room for that. Maps are future-proof.

Fictional example:
```clojure
{:dispatch/retry-on-fail false
 :dispatch/may-dispatch-to-service-worker true
 :dispatch [:display-pop-up]}
```

**Comment by olivergeorge** [comment]
> reg-event-fx2
> @olivergeorge I'm unclear on your reg-event-fx2 proposal (duplicated below)
> 
> (reg-event-fx2 :token (comp set-something fetch-data dispatch-other))
> Are you suggesting a new registration function which sets up the {:db db :fx []} part automatically and then threads that through the supplied event handler (in your example created via comp?

Yes, that point on the input side.  We'd want to pass in {:db db :fx []} plus any cofx regstered (e.g. :event, ...).

The return value is the other consideration.  We want :db and :fx to be handled but since :cofx will be passed through (e.g. :event we need them to be ignored).

e.g.
```cljs
(re-frame.core/reg-event-fx ::identity identity)
(re-frame.core/dispatch [::identity])

Console error: re-frame: no handler registered for effect: :event . Ignoring.
```

Questions
* What downsides would this new reg-event style have compared with reg-event-fx?
* Is reg-event-fx compatible already?  No, because of console errors thrown.
* Could reg-event-fx be made compatible?  Yes, by not throwing that error or providing a flag to disable it.  A noop is fine.
* Could it be handled with a helper?  Yes, only need to pick out :db and :fx from handler return value.

```cljs
(defn db-fx [m] (select-keys m [:db :fx]))
(defn reg-event-fx2
  ([id handler] (reg-event-fx2 id nil handler))
  ([id interceptors handler] (reg-event-fx id interceptors (comp db-fx handler))))
```

**Comment by kirillsalykin** [comment]
@mike-thompson-day8 @green-coder probably you are right and a map is the way to go.
Thanks for the points.

**Comment by olivergeorge** [comment]
Thinking more about how we could define & differentiate top level :db / :fx handlers and things appearing under :fx key.  I think I can see a way to unknot things whichout making a mess.

Perhaps we differentiate `effect` and `fx` ideas.
* `effects` returned by event handlers describe how the world should change.
* Use `reg-effect` to register handlers which process the change. 🆕 
* Use `reg-fx` to implement side effecting logic in your app.

This seems to align well with the current code base.  We need to add two new "effect" handlers for :db and :fx.

```
(reg-effect :fx
  (fn [vs]
    (doseq [[id & args :as v] vs :when v]
      (if-let [f (get-handler :fx id true)]
        (apply f args)
        (console :warn (str "No registered fx handler for " id))))))

(reg-effect :db
  (fn [db]
    (reset! app-db db)))
```

What this effectively does is demote reg-fx.

(I see a few options to facilitate backward compatibility)

**Comment by mike-thompson-day8** [comment]
@olivergeorge 
I appreciate the idea, and I can see your point, but I'm also comfortable with just having one kind of effect handler registered via `reg-fx`. After that, it is a training/docs issue.

**Comment by mike-thompson-day8** [comment]
We're just pushing out a new release with this change in it.  See v1.1.0 

Still to be done:
  - [ ] tutorial documentation has to be slowly evolved towards use of `:fx`  (there effectively been a change in best practice) 
  - [ ] more discussion about:  should there be an alternative to `reg-event-db` and `reg-event-fx` which is better at allowing handlers to be created as a composition of functions?

**Comment by mike-thompson-day8** [comment]
Many Thanks for @olivergeorge for the idea.  I think this approach will mature nicely.

**Comment by tangjeff0** [comment]
This is awesome! One question: should `re-frame-async-flow-fx` be updated as well? It follows the `dispatch-n` pattern.

**Comment by danielytics** [comment]
At first I wasn't too sure what the value of the new approach is, for compostability (I mean, `assoc`ing to a map is just as easy as `conj`ing to a vector) but after converting some of my code to the new way, the value is very apparent and I love this approach. Not having to think about dispatch vs dispatch later or otherwise checking if an effect is already present makes for some very neatly composable code! I'm really liking how clean its making my handlers. Great work :+1:

**Comment by mike-thompson-day8** [comment]
@olivergeorge 

As I understand it, this is effectively what you'd like to see.  Except you'd like to see it made easier via a new `reg-event-xxx`. 
```
(reg-event-ctx
  :event-id
  (fn [{:keys [coeffects] :as context}]
    (let [initial  {:db     (:db coeffects)
                       :event  (:event coeffects)
                       :fx     []}
           result   (-> initial
                         function1 
                         function2
                         function3)
            effects  (selectkeys result [:db :fx])]
      (assoc context :effects effects))))
```

**Comment by olivergeorge** [comment]
@mike-thompson-day8 that reads fine to me.

One suggestion, perhaps `initial` should be `coeffects` to allow for handler specific interceptors to inject cofx and have them available to the composed handlers.

**Comment by mike-thompson-day8** [comment]
@olivergeorge   I just edited my suggestion above to include "Problems".

**Comment by mike-thompson-day8** [comment]
@olivergeorge 

Ah, using your idea and adding another, I've got it.  All problems solved ...

Schematic:
```clj
(reg-event-ctx      ;; <-- I'm using the most basic wrapper to allow us to get clear on what's required
  :event-id
  (fn [{:keys [coeffects] :as context}]
    (let [initial  (assoc coeffects :fx [] :effects-keys [:db :fx] )     ;; <-- step 1

          result   (-> initial      ;; <-- step 2 
                     function1      ;; <-- composing the event handler from smaller functions
                     function2
                     function3)

          effects  (selectkeys result (:effects-keys result)]    ;; <-- step 3 - extract the effects
      (assoc context :effects effects))))            ;; <--  also step 3
```

So this formulation has three steps:

  1. Setup of `initial` (a data structure to thread through the functions) 
  2. The composition of many smaller functions which thread `initial`, progressively modifying it (typically just the `:db` and `:fx`).
  3. A way to extract just the effects out of the final threaded-through data 

Notes:
  1. `initial` is just the context's `coeffects` with two keys added:  `:fx` as an empty vector, and `effects-keys` which are the keys which should be later regarded as effects.  
  2. During step 2, functions can add to `:effect-keys` if they need to indicate that they have added another effect key (they typically won't need to)
  3. reframe will provide a new registration function: `reg-event-dbfx`  which will look after steps 1 and 3, allowing you to just provide the functions. 

## Outcome 

You'll be able to do this:
```clj 
(reg-event-dbfx  :event-id  interceptors [function1 function2 function3])
```
```

**Comment by olivergeorge** [comment]
Nice.  

Loving the explicit `:effect-keys`.

Not sure if the [fn1 fn2 fn2] form adds enough over (comp fn3 fn2 fn1) to justify itself.  Best argument I can see is that it does allow for the order of execution to be "left to right" which might be more intuitive.

**Comment by mike-thompson-day8** [comment]
We are about to implement and release `reg-event-dbfx`.  Any objections? Speak now or forever hold your peace.

**Comment by Ramblurr** [comment]
My only comment is one of naming `reg-event-dbfx`  seems confusing given existing re-frame functions. reg-event-db, reg-event-fx, and others. As a beginner/intermediate the names are so similar to seem overloaded. Just my 2 cents.

**Comment by mike-thompson-day8** [comment]
For the record, my earlier schematic using `reg-event-ctx` can be written using the less exotic `reg-event-fx` 

```clj
(reg-event-fx
  :event-id
  (fn [coeffects _]
    (let [initial  (assoc coeffects :fx [])    ;; <-- step 1

          result   (-> initial      ;; <-- step 2 
                     function1      ;; <-- composing the event handler from smaller functions
                     function2
                     function3)]

      (selectkeys result [:db :fx]))))    ;; <-- step 3 - extract and return the effects
```
(which can be written much more tersely) 

This realisation is encouraging me to wonder if we really need `reg-event-dbfx`.  Do we just need to teach the technique, not introduce a new registration mechanism?

**Comment by olivergeorge** [comment]
I can see the argument in waiting until there's evidence that the boilerplate is annoying or troublesome for third party libs which want to rely on composable behaviour.

**Comment by niquola** [comment]
Maps are better for long term design - less mental load; easy to compose and manipulate.  With ordered fx you force order, even if it is not needed.   Forcing order is one of sources of accidental complexity.  Order in functional langs are modeled with nesting - instead of switching to vector we can make all events and fxs to be maps and add special keys for chaining.  I can imagine collection of high order effects to compose fxs and events. Ideally multiple dispatches can be run in pure world as reduce with merge of fxs. @mike-thompson-day8 let's have a good reasonable complicated use case to exercise design ;)

**Comment by niquola** [comment]
About dispatch(-n) - are you sure it is effect at all?

**Comment by danielytics** [comment]
@niquola I, too, at first thought that maps are simpler and compose just as well, but I’ve since switched to this new way and my code really is simpler. The reason is that any function I call simply takes an fx map (`{:db ... :fx [...]}`) and returns an fx map, if they need to update the app state they run `(update-in fxmap [:db ...] ...)` and if they need to add an effect they run `(update fxmap :fx conj [new effect here])`.

The db update doesn’t change at all, but the effects now become much easier, because if I want to add an effect to be run, I simply add it to the vector. Order usually doesn’t matter so I can just conj it without a second thought (but if order did matter, it’s now an option too). This on first look doesn’t seem any better than just `assoc`ing a new effect into the old map, but consider this:

Before, each composed function needed to be aware of the existing effects in the map, or they might clobber them. Eg if you wanted to dispatch you always had to use dispatch-n in case there was already a dispatch effect (or check if it’s there). What if you want to add a http request? You had to check if there already was one in the map to not clobber it. Hopefully there was a http-n effect so you could add a second one. On that note, any time you wrote an effect, you had to provide an -n version, or make all your effects -n aware. This change makes that unnecessary and makes the effects vector explicit about what is being run. 

Personally, my code got simpler this way because I don’t need to think about what effects have already been added, I just conj to the end of the list. I have a function that sets up a http request and handles a response event. This sets up a number of dispatches and effects and the logic for this got much nicer with this change. If the dbfx sugar functions get added, I will likely be changing all of my handlers to using those.

I would suggest trying it out for any complex composed handlers you may have and see for yourself. If you don’t have any, it’s backwards compatible, so you can just stick with the old way. No need to change anything in that case.
